test/functional/mempool_cluster.py

```diff
@@ -0,0 +1,321 @@
+        utxo_to_spend = parent_tx["new_utxo"]
+        all_txids = [parent_tx["txid"]]
+        all_results = [parent_tx]
+
+        while len(all_results) < cluster_count:
+            next_tx = self.wallet.send_self_transfer(from_node=node, utxo_to_spend=utxo_to_spend, target_vsize=target_vsize)
+            assert next_tx["txid"] in node.getrawmempool()
+
+            # Confirm that each transaction is in the same cluster as the first.
+            assert_equal(node.getmempoolcluster(next_tx['txid']), node.getmempoolcluster(parent_tx['txid']))
+
+            # Confirm that the ancestors are what we expect
+            mempool_ancestors = node.getmempoolancestors(next_tx['txid'])
+            assert_equal(sorted(mempool_ancestors), sorted(all_txids))
+
+            # Confirm that each successive transaction is added as a descendant.
+            assert all([ next_tx["txid"] in node.getmempooldescendants(x) for x in all_txids ])
+
+            # Update for next iteration
+            all_results.append(next_tx)
+            all_txids.append(next_tx["txid"])
+            utxo_to_spend = next_tx["new_utxo"]
+
+        assert node.getmempoolcluster(parent_tx['txid'])['txcount'] == cluster_count
+        return all_results
+
+    def check_feerate_diagram(self, node):
+        """Sanity check the feerate diagram."""
+        feeratediagram = node.getmempoolfeeratediagram()
+        last_val = [0, 0]
+        for x in feeratediagram:
+            # The vsize is always positive, except for the first iteration
+            assert x['vsize'] > 0 or x['fee'] == 0
+            # Monotonically decreasing fee per vsize
+            assert_greater_than_or_equal(last_val[0] * x['fee'], last_val[1] * x['vsize'])
+            last_val = [x['vsize'], x['fee']]
+
+    def test_limit_enforcement(self, cluster_submitted, target_vsize_per_tx=None):
+        """
+        the cluster may change as a result of these transactions, so cluster_submitted is mutated accordingly
+        """
+        # Cluster has already been submitted and has at least 3 transactions, otherwise this test won't work.
+        assert_greater_than_or_equal(len(cluster_submitted), 3)
+        node = self.nodes[0]
+        last_result = cluster_submitted[-1]
+
+        # Test that adding one more transaction to the cluster will fail.
+        bad_tx = self.wallet.create_self_transfer(utxo_to_spend=last_result["new_utxo"], target_vsize=target_vsize_per_tx)
+        assert_raises_rpc_error(-26, "too-large-cluster", node.sendrawtransaction, bad_tx["hex"])
+
+        # It should also work during replacement
+        utxo_to_double_spend = self.wallet.get_utxo(confirmed_only=True)
+        fee = Decimal("0.000001")
+        tx_to_replace = self.wallet.create_self_transfer(utxo_to_spend=utxo_to_double_spend, fee=fee)
+        node.sendrawtransaction(tx_to_replace["hex"])
+
+        # Multiply fee by 5, which should easily cover the cost to replace. Otherwise, use the target vsize at 10sat/vB
```

Comment:
```suggestion
        # Multiply fee by 5, which should easily cover the cost to replace (but is still too large a cluster). Otherwise, use the target vsize at 10sat/vB
```